NR-228258 Logging too many files warning msg

Sivakumar3695 opened the pull request:
_Description:_

When there are too many files watched by the FB process, the OS will stop the process. To help users understand what went wrong, a warning log message will be produced which the user can take a look to understand the number of files targeted by each file paths provided for log forwarding.

With the changes in this PR, now, the Infra-Agent (while starting up) will compute the amount of target files for each file path config. If the total target files exceed the recommended maximum, the warning message will be printed in the logs.

jsubirat commented on pkg/integrations/v4/logs/cfg.go:
```diff
@@ -278,7 +299,11 @@ func NewFBConf(loggingCfgs LogsCfg, logFwdCfg *config.LogForward, entityGUID, ho
 
-	for _, block := range loggingCfgs {
+	totalFiles := 0
+	for i := 0; i < len(loggingCfgs); i++ {
```
You don't need to change the "for" shape here, simply include "i" now:
```
for i, block := range loggingCfgs {
// no need to do block := loggingCfgs[i] now
```

Sivakumar3695 replied:
Thanks for the suggestion! 🙂
I am not aware of this way!

jsubirat commented on pkg/integrations/v4/logs/cfg.go:
```diff
@@ -296,6 +321,18 @@ func NewFBConf(loggingCfgs LogsCfg, logFwdCfg *config.LogForward, entityGUID, ho
+			LogsCfg:    loggingCfgs,
+		}
+		warningMessage, err := warningDetails.Format()
+		if err != nil {
+			return FBCfg{}, err
+		}
```
We should not prevent the Infra-Agent to start if for some reason we have a formatting error here. Please do the following instead:
```
if err == nil {
    cfgLogger.Warn(warningMessage)
}
```

BTW, you can use the shorter syntax:
```
if warningMessage, err := warningDetails.Format(); err == nil {
    cfgLogger.Warn(warningMessage)
}
```

jsubirat commented on pkg/integrations/v4/logs/cfg_template.go:
```diff
@@ -152,3 +152,29 @@ var fbLuaScriptFormat = `function {{ .FnName }}(tag, timestamp, record)
+
+var tooManyFilesWarnMsg = `
+The amount of open files targeted by your Log Forwarding configuration
+files ({{ .TotalFiles }}) exceeds the recommended maximum (1024). The Operating System
```
Is there a way we can also pass the hardcoded maximum (1024) using the `fbFileWatchLimit` variable?

Sivakumar3695 replied:
Sure. I have added a field in the struct to avoid the hard coded recommended maximum.

jsubirat commented on pkg/integrations/v4/logs/cfg_template.go:
```diff
@@ -152,3 +152,29 @@ var fbLuaScriptFormat = `function {{ .FnName }}(tag, timestamp, record)
+These are the amount of files targeted by each of your configuration blocks:{{"\n"}}
+
+{{- range .LogsCfg }}
+{{- if .TargetFilesCnt }}
```
I'd say to check for non-null `. File` here instead, just in case there is some file path that contains no files. In this case, we'd still want to print a "0" for that path.

jsubirat commented on pkg/integrations/v4/logs/cfg_template.go:
```diff
@@ -152,3 +152,29 @@ var fbLuaScriptFormat = `function {{ .FnName }}(tag, timestamp, record)
+- You may also consider increasing the maximum amount of allowed file
+descriptors and inotify watchers. See: https://docs.newrelic.com/docs/logs/forward-logs/forward-your-logs-using-infrastructure-agent/#too-many-files
+
+If the maximum amount of allowed file descriptors and inotify watchers are increased already by following the above link,
+ignore this warning message.
```
```suggestion
Please note this is a friendly warning message. If your operating system allows more than 1024 file descriptors/inotify watchers or if you already increased their maximum amount by following the above link, you can safely ignore this message.
```

jsubirat replied:
WDYT?

Sivakumar3695 replied:
This should be good enough.

jsubirat commented on pkg/integrations/v4/logs/cfg_test.go:
```diff
@@ -1813,3 +1813,247 @@ func TestDetermineUseAnsiFlagValue(t *testing.T) {
+		name          string
+		logCfg        LogCfg
+		expectedCount int
+	}{
```
Can you please add test cases with:
- One that contains a glob of an existing folder but with no matches (should produce a 0 in the output message). i.e. `./*_test.starwars`
- One that contains a glob of a non-existing folder (should throw an error due to not having access to that folder or not existing): `./starwars/*_test.go`

Sivakumar3695 replied:
I have added a test case for the first one -> with no matches, will produce 0 target files

For the second test case suggested, the current code consumes the error inside the method itself. Hence, that cannot be tested in the unit testing. Instead, I have added a unit test case to check if 0 is returned as Target files count. I hope this is fine. Please feel free to improve or provide alternative solution.

jsubirat commented on pkg/integrations/v4/logs/cfg_test.go:
```diff
@@ -1813,3 +1813,247 @@ func TestDetermineUseAnsiFlagValue(t *testing.T) {
+						Name:           "log-file-2",
+						File:           "file.path-2",
+						TargetFilesCnt: 500,
+					},
```
Please also include a path with 0 files to stress what I mentioned in `cfg_template.go`:
```

{
    Name:           "log-file-3",
    File:           "file.path-3",
    TargetFilesCnt: 0,
}
```

jsubirat commented on pkg/integrations/v4/logs/cfg_test.go:
```diff
@@ -1813,3 +1813,247 @@ func TestDetermineUseAnsiFlagValue(t *testing.T) {
+`,
+		},
+		{
+			"Too Many files with Syslog, Tcplog, ystemdlog, winlog, wineventlog",
```
Nice one! 🙇

jsubirat reviewed (commented):
Really nice work @Sivakumar3695 ! I left you some comments for you to consider. 🙇

Sivakumar3695 commented on pkg/integrations/v4/logs/cfg.go:
```diff
@@ -317,6 +354,18 @@ func NewFBConf(loggingCfgs LogsCfg, logFwdCfg *config.LogForward, entityGUID, ho
+	}
+	files, err := filepath.Glob(l.File)
+	if err != nil {
+		cfgLogger.WithField("filePath", l.File).Error("Error while reading files under the given file path")
```
I have included the error in the logger as below:
`cfgLogger.WithField("filePath", l.File).WithError(err).Error("Error while reading files under the given file path")
`

Sivakumar3695 commented:
Please take a look at the updated warning message
<img width="1496" alt="Screenshot 2024-03-04 at 5 37 21 PM" src="https://github.com/newrelic/infrastructure-agent/assets/29046579/d0f82813-449d-4b52-a6b8-5c6ca6a694e3">

Also, note the error message while reading the non-existing folder and non-matching file path.

Thank you!

jsubirat commented on pkg/integrations/v4/logs/cfg_template.go:
```diff
@@ -175,6 +175,6 @@ for the rotated logs (i.e. my_log.log.20240214 instead of my_log.20240214.log)
 
-If the maximum amount of allowed file descriptors and inotify watchers are increased already by following the above link,
-ignore this warning message.
+Please note that this is a friendly warning message. If your operating system allows more than {{ .DefaultFileLimit }} file descriptors/inotify watchers 
+or if you already increased their maximum amount by following the above link, you can safely ignore this message.
```
```suggestion
Please note that this is a friendly warning message. You can safely ignore this message if your operating system allows more than {{ .DefaultFileLimit }} file descriptors/inotify watchers 
or if you have already increased their maximum amount by following the above link.
```

Sivakumar3695 replied:
Yeah. This would be much more opt than the previous message. I agree with you on this one and have update the message accordingly. Nice suggestion @jsubirat

jsubirat reviewed (approved):
Nice work, thanks a lot Siva! I have made up my mind again regarding the displayed message, feel free to ignore the suggestion if you think it's already alright.

rogercoll reviewed (commented):
The code LGTM, but I have a minor concern regarding the amount of lines this warning can generate. Actually, the infastructure-agent does not have any multiline log entry and the format can be changed to "json" instead of plain text. Could it happen that the logging directory has many subdirectories? If so, the warning text would be quite long. Is it required to show detailed information of each directory? Could we move the details of each subdirectory to trace level?

rogercoll commented:
@jsubirat @Sivakumar3695 Currently, the Fluent-bit infrastructure agent's supervisor auto removes temporal configuration files if there are more than 50: https://github.com/newrelic/infrastructure-agent/blob/master/pkg/integrations/v4/supervisor_fb.go#L32

Should the fb supervisor auto remove this files too?

rogercoll commented:
Related to the generated warning message, would it be correctly parsed by fluent-bit itself? Would it take into account the multiline text? Have you considered pointing to public documentation URL or local file path instead?

Sivakumar3695 commented:
Hi @rogercoll 
Thanks for your feedback. Here are my thoughts on your feedback comments:

> The code LGTM, but I have a minor concern regarding the amount of lines this warning can generate. Actually, the infastructure-agent does not have any multiline log entry and the format can be changed to "json" instead of plain text. Could it happen that the logging directory has many subdirectories? If so, the warning text would be quite long. Is it required to show detailed information of each directory? Could we move the details of each subdirectory to trace level?

I will try to work on this and will come up with an alternative way to convey the same message. Moving to Json? Idk, probably, I need to check to confirm this. Currently, we print the generated fluent-bit config file in debug and I will look if the same can be done here as well.

Sivakumar3695 commented:
> Currently, the Fluent-bit infrastructure agent's supervisor auto removes temporal configuration files if there are more than 50: https://github.com/newrelic/infrastructure-agent/blob/master/pkg/integrations/v4/supervisor_fb.go#L32. Should the fb supervisor auto remove this files too?

No. These are log files of the users that they want to ship the log data to newrelic. This is not something that can't be removed by the Infra-agent. We provide the warning message to inform the user that "Too many files are opened for log forwarding" and they need to increase the number of files that can be opened by a process in an OS.

Sivakumar3695 commented:
> Related to the generated warning message, would it be correctly parsed by fluent-bit itself? Would it take into account the multiline text? Have you considered pointing to public documentation URL or local file path instead?

I will check once on whether the multi-line warning message can be correctly processed by fluent-bit. For the public doc link suggested by you, in fact, we provided a public documentation URL in the warning message itself. However, the detailed warning message is only to help users / a GTSE engineer to understand the root cause of the issue by giving insights into the folders containing too many files.

Sivakumar3695 commented:
@jsubirat anything that you would like to add here?

Sivakumar3695 commented:
@rogercoll Would it be fine to move the warning message to a log field as below?

> cfgLogger.WithField("tooManyFilesWarning", warningMessage).Warn(
				"The amount of open files targeted by your Log Forwarding configuration " +
					"files exceeds the recommended maximum.")

This way, the multi-line warning message in discussion would be wrapped inside a log field similar to how we log fluent-bit configuration file [here](https://github.com/newrelic/infrastructure-agent/blob/ec2d9ffc325e1c3d9388fa586f897d561c74e5dc/pkg/integrations/v4/supervisor_fb.go#L249).

@jsubirat We can reduce the log level for this to **Trace**. WDYT?

Sivakumar3695 commented:
Hi @rogercoll. 
After our discussion, I have updated the warning message as below:
1. Now, we have a non-multiline warning log message with log level -> Warn
<img width="1446" alt="Screenshot 2024-03-13 at 5 48 22 PM" src="https://github.com/newrelic/infrastructure-agent/assets/29046579/7939784c-5d2a-4703-b38e-3b8defe2908a">

2. We moved the detailed information of each directory to individual log message in Trace level
<img width="1452" alt="Screenshot 2024-03-13 at 5 48 11 PM" src="https://github.com/newrelic/infrastructure-agent/assets/29046579/7654641e-a02e-4a41-ac8b-c80ba8078660">

rogercoll commented on pkg/integrations/v4/logs/cfg.go:
```diff
@@ -88,17 +89,18 @@ type YAML struct {
+	Fluentbit      *LogExternalFBCfg `yaml:"fluentbit"`
+	Winlog         *LogWinlogCfg     `yaml:"winlog"`
+	Winevtlog      *LogWinevtlogCfg  `yaml:"winevtlog"`
+	TargetFilesCnt int
```
Small nit, can we make this new field not public? It should not be used outside this package

rogercoll reviewed (commented):
Thank you @Sivakumar3695 ! New code changes look good to me, just added a small comment

rogercoll commented on pkg/integrations/v4/logs/cfg_test.go:
```diff
@@ -1813,3 +1813,117 @@ func TestDetermineUseAnsiFlagValue(t *testing.T) {
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			NewFBConf(test.logsCfg, &test.logFwd, "0", "")
+			assert.Equal(t, test.expectedCount, test.logsCfg[0].TargetFilesCnt)
```
I reckon this won't compile as the `targetFilesCnt` is not exported anymore

rogercoll reviewed (approved):
LGTM
